pkg/validate/validate_test.go

```diff
@@ -0,0 +1,304 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/ghodss/yaml"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/kylelemons/godebug/diff"
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	"istio.io/operator/pkg/util"
+)
+
+func TestUnmarshalKubernetes(t *testing.T) {
+	tests := []struct {
+		desc    string
+		yamlStr string
+		want    string
+	}{
+		{
+			desc:    "nil success",
+			yamlStr: "",
+			want:    "{}",
+		},
+		{
+			desc: "hpaSpec",
+			yamlStr: `
+hpaSpec:
+  maxReplicas: 10
+  minReplicas: 1
+  targetCPUUtilizationPercentage: 80
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: php-apache
+`,
+		},
+		{
+			desc: "resources",
+			yamlStr: `
+resources:
+  limits:
+    cpu: 444m
+    memory: 333Mi
+  requests:
+    cpu: 222m
+    memory: 111Mi
+`,
+		},
+		{
+			desc: "podDisruptionBudget",
+			yamlStr: `
+podDisruptionBudget:
+  maxUnavailable: 1
+  selector:
+    matchLabels:
+      app: pilot
+`,
+		},
+		{
+			desc: "readinessProbe",
+			yamlStr: `
+readinessProbe:
+  failureThreshold: 44
+  initialDelaySeconds: 11
+  periodSeconds: 22
+  successThreshold: 33
+  handler: {}
+`,
+		},
+		{
+			desc: "affinity",
+			yamlStr: `
+affinity:
+  podAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+    - labelSelector:
+        matchExpressions:
+        - key: security
+          operator: In
+          values:
+          - S1
+      topologyKey: failure-domain.beta.kubernetes.io/zone
+  podAntiAffinity:
+    preferredDuringSchedulingIgnoredDuringExecution:
+    - weight: 100
+      podAffinityTerm:
+        labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S2
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+`,
+		},
+		{
+			desc: "k8sObjectOverlay",
+			yamlStr: `
+overlays:
+- kind: Deployment
+  name: istio-citadel
+  patches:
+  - path: spec.template.spec.containers.name:galley.ports.containerPort:15014
+    value: 12345
+`,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			tk := &v1alpha2.TestKube{}
+			err := unmarshalWithJSONPB(tt.yamlStr, tk)
+			if err != nil {
+				t.Fatalf("unmarshalWithJSONPB(%s): got error %s", tt.desc, err)
+			}
+			s, err := marshalWithJSONPB(tk)
+			if err != nil {
+				t.Fatalf("unmarshalWithJSONPB(%s): got error %s", tt.desc, err)
+			}
+			got, want := stripNL(s), stripNL(tt.want)
+			if want == "" {
+				want = stripNL(tt.yamlStr)
+			}
+			if !util.IsYAMLEqual(got, want) {
```

Comment:
I couldn't find `IsYAMLEqual` in util, but there is a similar func in `values_test.go` - that should probably be moved to `util`

Reply:
Again, this function has moved. I've updated here: https://github.com/istio/operator/pull/40